### DISCUSSION TITLE
Comparison of mjd_transitionFD on simple inverted pendulum model with MATLAB c2d

### DISCUSSION THREAD

**Discussion by dtorre38** [google-deepmind/mujoco#707] [General]
Hello,

I am interested in using mjd_transitionFD for my simulations, but I wanted to see how it compared with MATLAB's c2d. So I ran mjd_transitionFD and c2d on a simple inverted pendulum model. I used several of c2d's discretization methods and have tabulated the results below:

![Screenshot 2023-01-24 at 12 05 15](https://user-images.githubusercontent.com/84144166/214372927-0fba221b-eab7-4346-995f-8d40948aceb6.png)

Math is according to [https://en.wikipedia.org/wiki/Discretization](url)

In comparison to MATLAB's ZOH method, mjd_transitionFD has some differences in the A matrix (A21) and the entire B matrix. I know that the MuJoCo API mentions that mjd_transitionFD produces "Finite differenced state-transition and control-transition matrices dx(t+h) = A*dx(t) + B*du(t)". I have also reviewed the humanoid standing balance, so I know that the A and B matrices which are produced are good enough for control, but if possible I wanted more insight into the inner workings of mjd_transitionFD. 

-Daniel

**Comment by yuvaltassa** [comment]
1. I don't understand how you made this comparison. You're using a Matlab wrapper for MuJoCo, or you wrote your own dynamics and are hoping that they match MuJoCo's? My guess is that you did the latter and that there is a mismatch between the models.
2. The inner workings are [right here](https://github.com/deepmind/mujoco/blob/cd99fbd59dc75fa8be97fe90a391d2055ee0a4d6/src/engine/engine_derivative.c#L1725).

**Reply by dtorre38** [reply]
I simply wrote my own dynamics and obtained all the values from the MJMODEL.txt file. Thank you for posting the link to the mjd_transitionFD code!

**Reply by yuvaltassa** [reply]
I would guess that your integrator is different from MuJoCo. The [Euler integrator](https://github.com/deepmind/mujoco/blob/main/src/engine/engine_forward.c#L545) explained [here](https://mujoco.readthedocs.io/en/latest/computation.html#numerical-integration), is not simple Euler. It is semi-implicit and has a fully implicit component (w.r.t joint damping).

**Reply by dtorre38** [reply]
I only ran mjd_transitionFD when initializing the controller in init_controller(), not for every integration time step. Similarly in MATLAB I did not run a simulation, I only converted the dynamics from  continuous to discrete. For both I assumed the upright unstable fixed point. I wanted the initial set of values to get a sense of how the linearization performed in comparison to c2d.
